lib/AST/ASTImporter.cpp

```diff
@@ -4392,31 +4392,45 @@ Decl *ASTNodeImporter::VisitClassTemplateDecl(ClassTemplateDecl *D) {
     for (unsigned I = 0, N = FoundDecls.size(); I != N; ++I) {
       if (!FoundDecls[I]->isInIdentifierNamespace(Decl::IDNS_Ordinary))
         continue;
-      
+
       Decl *Found = FoundDecls[I];
-      if (ClassTemplateDecl *FoundTemplate 
-                                        = dyn_cast<ClassTemplateDecl>(Found)) {
+      if (ClassTemplateDecl *FoundTemplate =
+              dyn_cast<ClassTemplateDecl>(Found)) {
+
+        // We found a forward declaration but the class to be imported has a
+        // definition.
+        if (D->isThisDeclarationADefinition() &&
+            !FoundTemplate->isThisDeclarationADefinition()) {
+
+          // Lookup will find the fwd decl only if that is more recent than the
+          // definition. So, lets try to get the definition if that available
+          // in the redecl chain.
+          CXXRecordDecl *ToTemplatedDef =
+              FoundTemplate->getTemplatedDecl()->getDefinition();
+          if (!ToTemplatedDef)
+            continue;
+          ClassTemplateDecl *TemplateWithDef =
+              ToTemplatedDef->getDescribedClassTemplate();
+          if (!TemplateWithDef)
```

Comment:
Can this be triggered? Should this be an assert?

Reply:
Yes. In case of ClassTemplate import itself. Here are the steps:
- we import the templated CXXRecordDecl (from VisitClassTemplateDecl)
- then we start the import of the ClassTemplateDecl again (from VisitRecordDecl)
- during that second call to VisitClassTemplateDecl the described class is not set already, we set that right after the creation of the ClassTemplateDecl:
```
  bool AlreadyImported;
  ClassTemplateDecl *D2;
  std::tie(D2, AlreadyImported) = CreateDecl<ClassTemplateDecl>(
      D, Importer.getToContext(), DC, Loc, Name, TemplateParams, ToTemplated);
  if (AlreadyImported)
    return D2;

  ToTemplated->setDescribedClassTemplate(D2);
```